File: capgemini/batch-3/08-02-24/Kadanes.py



def getMaximumSubarraySum(arr):
    ans = float("-inf")
    currSum = 0
    for i in range(len(arr)):
        if currSum <=0:
            currSum = 0
        if arr[i]>=0:
            currSum+=arr[i]
        if arr[i]<0:
            ans = max(ans,currSum)
            currSum+=arr[i]
    return max(ans,currSum)


def main():
    arr = [-3,2,5,-7,9,12,-3,-5,6,8,12]
    ans = getMaximumSubarraySum(arr)
    print(ans)
main()

File: madurai/S22/26-06-2024/Kadanes.java
public class Kadanes {
    public static void main(String[] args) {
        int[] arr = {-3,4,2,5,-5,-7,10,4,-89};
        int ans = getMaximumSubarraySum(arr);
        System.out.println(ans);
    }

    private static int getMaximumSubarraySum(int[] arr) {
       int ans = Integer.MIN_VALUE;
       int cts = 0;

       for(int i=0;i<arr.length;i++){
            if(cts < 0) cts = 0;
            cts += arr[i];
            ans = Math.max(ans,cts);

       }
       return Math.max(ans,cts);
    }
}
